engine/component.ftl

```diff
@@ -206,6 +206,94 @@
+    [#-- If the value has been prefixed with the shared attribute then the prefix will be stripped and provider attribtue omitted --]
+    [#-- If the default value has been added on the provider attribute and the value is a string then the provider is prefixed --]
+    [#-- if the default has been prefixed with the shared attribute then the prefix will be stripped and the provider attribute omitted --]
+    [#-- When handling a list of values the shared and provider attributes need to be specified --]
+
+    [#local attributes = []]
+    [#list providerAttributes as providerAttribute ]
+        [#local prefixedNames = [] ]
+        [#local prefixedValues = [] ]
+        [#local prefixedDefault = "" ]
+        [#list asArray(providerAttribute.Names!providerAttribute.Name![]) as name]
+
+            [#-- This is to allow for providers to extend existing attributes without having them prefixed --]
+            [#if name?starts_with("${SHARED_ATTRIBUTES}:")]
+
+                [#local prefixedNames += [ name?remove_beginning("${SHARED_ATTRIBUTES}:") ]]
+
+                [#-- If the attribute is a shared attribute prefix the potential values and defaults to make them unique --]
+                [#if ((providerAttribute.Values)![])?has_content ]
+
+                    [#local prefixedValues = providerAttribute.Values?map(
+                            x -> x?is_string?then(
+                                                x?starts_with("${SHARED_ATTRIBUTES}:")?then(
+                                                    x?remove_beginning("${SHARED_ATTRIBUTES}:"),
+                                                    x?ensure_starts_with("${provider}:" )
+                                                ),
+                                                x)
+                            )]
+                [/#if]
+
+                [#if ((providerAttribute.Default)!"")?has_content ]
+                    [#if (providerAttribute.Default)?is_sequence ]
+                        [#local prefixedDefault = providerAttribute.Default?map(
+                                x -> x?is_string?then(
+                                    x?ensure_starts_with("${provider}:"),
+                                    x)
+                                )]
+                    [#elseif (providerAttribute.Default)?is_string ]
+                        [#local prefixedDefault = providerAttribute.Default?ensure_starts_with("${provider}:") ]
+                    [#else]
+                        [#local prefixedDefault = providerAttribute.Default ]
+                    [/#if]
+                [/#if]
+
+            [#else]
+
+                [#local prefixedNames += [ name?ensure_starts_with("${provider}:")]]
+
+                [#local prefixedValues = (providerAttribute.Values?map( x -> x?is_string?then( x?remove_beginning("${SHARED_ATTRIBUTES}:"), x ) ))![] ]
```

Comment:
So if a provider specific name, should all the values be forced to be prefixed with the provider or is the assumptions that the values are unprefixed because the attribute is provider specific?

Also wondering what the intent of removing any "shared:" on the front was - when would this be provided? Or is this just ensuring any occurrence of the "shared:" prefix is always stripped off reagrdless?

Reply:
anything without the `shared:` prefix will be prefixed with the provider prefix. If you append an attribute with shared your essentially saying I want to use attributes from the shared provider but include my own. 

On further review of our code @RossMurr4y has already implemented his behaviour in extensions. So most of this code isn't actually required. I'm working on some updates in the extensions which will probably replace this PR and use extensions instead